### ISSUE TITLE
graceful shutdown order and cleanup contexts

### ISSUE THREAD

**Issue by hhromic** [fafhrd91/aiohttp#8448]
### Describe the bug

The current graceful shutdown order in aiohttp does not consider calling the second stage, i.e. any code after `yield`, of cleanup contexts before waiting for any remaining/pending left-over tasks to complete. This makes the server to take a long time (up to `shutdown_timeout`) if cleanup contexts create their own tasks during the life-time of the server after starting because any task cancellation after `yield` is called too late.

The current order as described [in the documentation](https://docs.aiohttp.org/en/stable/web_advanced.html#graceful-shutdown) is:
1. Stop each site listening on sockets, so new connections will be rejected.

2. Close idle keep-alive connections (and set active ones to close upon completion).

3. Call the [Application.on_shutdown](https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.Application.on_shutdown) signal. This should be used to shutdown long-lived connections, such as websockets (see below).

4. Wait a short time for running tasks to complete. This allows any pending handlers or background tasks to complete successfully. The timeout can be adjusted with shutdown_timeout in [run_app()](https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.run_app).

5. Close any remaining connections and cancel their handlers. It will wait on the canceling handlers for a short time, again adjustable with shutdown_timeout.

6. Call the [Application.on_cleanup](https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.Application.on_cleanup) signal. This should be used to cleanup any resources (such as DB connections). This includes completing the [cleanup contexts](https://docs.aiohttp.org/en/stable/web_advanced.html#aiohttp-web-cleanup-ctx).

7. Cancel any remaining tasks and wait on them to complete.

With this order, completing cleanup contexts (6) is performed _after_ waiting for remaing running tasks to complete (4). Even if the cleanup contexts are designed to cancel/stop their own tasks after `yield`.

In my humble opinion, (6) should be performed before (3) so cleanup contexts are completed and `on_cleanup()` signals are called shortly before `on_shutdown()` signals. I'm not sure why cleanup contexts and `on_cleanup()` signals are delayed until the very end in contrast to `on_shutdown()`. Another option could be placing (4) after (6).

I discovered this issue because our aiohttp server applications include an aiokafka client using a cleanup context. When request handlers send data to Kafka, aiokafka creates background tasks to batch and send the actual data, which are normally stopped after `yield` in the cleanup context. However this is not called early enough and every time we terminate the server, it waits until `shutdown_timeout` (default 60s) before exiting cleanly.

The reproduction example provided below mimics exaclt what happens when using aiokafka.

A very similar issue was opened before here: <https://github.com/aio-libs/aiohttp/issues/8179>
In that issue, it was even [suggested](https://github.com/aio-libs/aiohttp/issues/8179#issuecomment-1992129569) to use cleanup contexts or separate startup/shutdown (not cleanup?) signals.

### To Reproduce

Reproduction code:
```python
import asyncio
from aiohttp import web

class Scheduler:
    def __init__(self):
        self.tasks = []

    async def cancel(self):
        any(t.cancel() for t in self.tasks)
        await asyncio.gather(*self.tasks)

    async def schedule(self):
        async def _worker():
            for i in range(30):
                print(f"beep bop {i}")
                await asyncio.sleep(1)
        self.tasks += [asyncio.create_task(_worker())]

async def scheduler_ctx(app):
    app["scheduler"] = Scheduler()
    yield
    await app["scheduler"].cancel()

async def hello(request):
    await request.app["scheduler"].schedule()
    return web.Response(text="Hello, world")

app = web.Application()
app.add_routes([web.get("/", hello)])
app.cleanup_ctx.append(scheduler_ctx)

web.run_app(app)
```

1. Run the above example.
2. Terminate the server with CTRL-c and observe it immediately shuts down cleanly.
3. Re-run the above example again.
4. Do a GET request to `/` to trigger the handler and create a scheduler dummy worker task.
5. Try to terminate the server with CTRL-c and observe that the second stage in the cleanup context (the call to `Scheduler.cancel()` is _not_ called before the shutdown timeout expires, and therefore the server will wait until the dummy worker finishes. If the dummy worker is configured to last more than 60 seconds (the default shutdown timeout), then the server will terminate at 60 seconds.

### Expected behavior

The expected behaviour is that the second stage of all defined cleanup contexts is called _before_ aiohttp waits for any pending/remaining unmonitored tasks to be completed during a graceful shutdown.

Cleanup contexts are meant to manage the life-cycle of background services in the same manner as `on_startup()` and `on_cleanup()` signals do, but combined together. Ref: <https://docs.aiohttp.org/en/stable/web_advanced.html#cleanup-context>

### Logs/tracebacks

```python-traceback
$ python server.py
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
beep bop 0
beep bop 1
beep bop 2
beep bop 3
beep bop 4
beep bop 5
beep bop 6
beep bop 7
beep bop 8
beep bop 9
beep bop 10
beep bop 11
beep bop 12
beep bop 13
beep bop 14
^Cbeep bop 15
beep bop 16
beep bop 17
beep bop 18
beep bop 19
beep bop 20
beep bop 21
beep bop 22
beep bop 23
beep bop 24
beep bop 25
beep bop 26
beep bop 27
beep bop 28
beep bop 29
```

### Python Version

```console
$ python --version
Python 3.12.3
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.9.5
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Location: /home/hhromic/.conda/envs/py312/lib/python3.12/site-packages
Requires: aiosignal, attrs, frozenlist, multidict, yarl
Required-by:
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.5
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/hhromic/.conda/envs/py312/lib/python3.12/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.4
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /home/hhromic/.conda/envs/py312/lib/python3.12/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

Debian Linux under WSL2

### Related component

Server

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

**Comment by Dreamsorcerer** [comment]
This is working as intended. If you want to cancel the tasks first, then use on_shutdown (maybe we could even add a shutdown_ctx or something...).

In the typical case, such a scheduler may be spawning background tasks for DB writes or similar. Such tasks may all complete within a second or two and so it's preferable to give them a chance to complete, rather than cancelling them immediately. You might actually end up with 2 different schedulers for the different use cases, one on shutdown and one on cleanup.

> The expected behaviour is that the second stage of all defined cleanup contexts is called before aiohttp waits for any pending/remaining unmonitored tasks to be completed during a graceful shutdown.

If you think about the meaning of 'cleanup', this typically refers to cleaning up resources, such as database connections. If we were to do this _before_ waiting on tasks, then the tasks we're waiting on would likely break as the resources they need have just been removed.

**Comment by hhromic** [comment]
Hi @Dreamsorcerer , thanks for the quick reply.
Perhaps the wording in my reproduction example was not the best, apologies!

While I do _cancel_ the tasks in the reproduction example, in the case of aiokafka we actually call a `Client.stop()` method which in itself also performs graceful shutdown of all the running background tasks of the Kafka client. This means that they are not abruptly canceled in the cleanup context.

Perhaps your suggestion of adding a `shutdown_ctx` for startup/shutdown pairs similar to how `cleanup_ctx` exists for startup/cleanup pairs could be useful indeed and make life-cycle management more flexible.

For reference, the following is a basic example using a real aiokafka producer client:
```python
import asyncio
from aiohttp import web
from aiokafka import AIOKafkaProducer

async def kafka_producer(app):
    app["kafka_producer"] = AIOKafkaProducer(
        bootstrap_servers="localhost:9092",
    )
    yield
    await app["kafka_producer"].stop()

async def hello(request):
    await request.app["kafka_producer"].send("some-topic", b"Hello, world")
    return web.Response(text="Hello, world")

app = web.Application()
app.add_routes([web.get("/", hello)])
app.cleanup_ctx.append(kafka_producer)  # or shutdown_ctx ?

web.run_app(app)
```

For this example to work correctly (no shutdown timeout), it would need to use the proposed `shutdown_ctx` because we need to ask the aiokafka client to perform a graceful shutdown before aiohttp does the wait-then-cleanup section.

My last concern would be then how we distinguish what goes into `cleanup_ctx` or `shutdown_ctx`?
The current documentation about shutdown/cleanup states the following:

* `on_shutdown`
   A Signal that is fired on application shutdown.
   Subscribers may use the signal for gracefully closing long running connections, e.g. websockets and data streaming.
* `on_cleanup`
   A Signal that is fired on application cleanup.
   Subscribers may use the signal for gracefully closing connections to database server etc.

The aiokafka client classifies more as a "database server" than "long running connections", but we have established that is better suited for its life-cycle to be managed using `on_shutdown`.

Thanks for your patience and looking forward on improving the usability of aiohttp in this regard!

**Comment by Dreamsorcerer** [comment]
I think in this case, you do want it to be on_cleanup. Otherwise, there could be a chance that a request handler for hello() might try to send a message _after_ the producer has been stopped.

If the problem is that aiokafka is creating new tasks after startup that are meant to sit idle, then I think this is a duplicate of #8387. In this case, I'm starting to think that we should actually track the tasks created for response handlers, and only wait on those tasks to complete. Then make the documentation clear that background tasks that should be allowed to complete should be scheduled through aiojobs and waited on in the cleanup_ctx explicitly.

**Comment by hhromic** [comment]
> If the problem is that aiokafka is creating new tasks after startup that are meant to sit idle, then I think this is a duplicate of #8387.

Yes, basically for the case of aiokafka is exactly that it creates its own tasks in the background after startup. But I think this problem can be generalised to anything that does the same. The issue you linked has the example for a postgres client. And yes I agree this issue is a duplicate of that one. Thanks for linking as I didn't see it before.

> In this case, I'm starting to think that we should actually track the tasks created for response handlers, and only wait on those tasks to complete.

I was looking into the code of aiohttp too to see how to improve on this and I was reaching the same conclusion indeed. Aiohttp could track its own created tasks and only wait on those. This would fully solve my issue and #8387, and likely be more robust because aiohttp would not need to do assumptions on the tasks submitted to the event loop.

Given that this is indeed a duplicate, I'm going to close this issue in favour of #8387 and follow that issue instead.

In the meantime, while it is not resolved, our current workaround is to simply configure a way shorter `shutdown_timeout`, e.g. 5s, so the hang is not that long on termination.

**Comment by Dreamsorcerer** [comment]
> Aiohttp could track its own created tasks and only wait on those.

So, the original reason for this design is so we'd wait on all user-created tasks. For example, if you use asyncio.shield() to protect a database write, this will then be a task which can't be tracked by aiohttp. So, the idea was to wait on all these things. It seems the approach is conflicting with too many other libraries though, so I guess we'll just have to rely on documentation to encourage users to ensure such tasks are tracked by their own code.

**Comment by Dreamsorcerer** [comment]
> I was looking into the code of aiohttp too to see how to improve on this and I was reaching the same conclusion indeed.

If you'd like to have a go at implementing this change, that'd be very helpful.

**Comment by hhromic** [comment]
> If you'd like to have a go at implementing this change, that'd be very helpful.

Thanks for the offer, but I think I'm not versed enough on aiohttp's codebase for such a change :(

I'll try to take a look but can't promise any progress, at least not any time soon 🙈
